prime-router/src/main/kotlin/azure/EmailEngineFunction.kt

```diff
@@ -0,0 +1,341 @@
+    @StorageAccount("AzureWebJobsStorage")
+    fun deleteEmailSchedule(
+        @HttpTrigger(
+            name = "deleteEmailSchedule",
+            methods = [HttpMethod.DELETE],
+            authLevel = AuthorizationLevel.ANONYMOUS,
+            route = "email-schedule/{scheduleId}"
+        ) request: HttpRequestMessage<String?>,
+        @BindingName("scheduleId") scheduleId: Int,
+        context: ExecutionContext,
+    ): HttpResponseMessage {
+        var user:String? = validateUser( request, context.logger );
+        var ret = request.createResponseBuilder(HttpStatus.UNAUTHORIZED);
+
+        if( user !== null ){
+            val id = WorkflowEngine.databaseAccess.deleteEmailSchedule(scheduleId);
+            ret.status(HttpStatus.CREATED)
+            ret.body( "deactivated id $id")
+        }
+        return ret.build();
+    }
+
+    /**
+     * Timer Trigger to fire when processing schedules
+     *   (fires every 5 minutes)
+     */
+    @FunctionName("emailScheduleEngine")
+    @StorageAccount("AzureWebJobsStorage")
+    fun run(
+        @TimerTrigger( name = "emailScheduleEngine", schedule = SCHEDULE ) timerInfo : String,
+        context: ExecutionContext
+    ){
+        val logger : Logger = context.logger;
+        val mapper = ObjectMapper().registerModule(KotlinModule());
+
+        // get the schedules to fire
+        val schedulesToFire : List<EmailSchedule> = WorkflowEngine.databaseAccess.fetchEmailSchedules()
+            .map{ mapper.readValue<EmailSchedule>( it ) }
+            .filter{ shouldFire( it ) }
+
+        schedulesToFire.forEach { schedule -> 
+            val orgs: Iterable<String> = getOrgs( schedule );
+
+            logger.info( "processing ${schedule.template}" )
+
+            // get the orgs to fire for
+            orgs.forEach{ org ->
+                    val emails: List<String> = if ( schedule.emails.size > 0) schedule.emails else getEmails(org, logger)
+                    logger.info( "processing ${org}" )
+                    emails.forEach{ email -> 
+                        logger.info( "sending email to ${email}" )
+                        dispatchToSendGrid( schedule.template, listOf(email), logger );
+                    }
+            }
+        }
+    }
+
+    /**
+     * Determine if a schedule should fire
+     * 
+     * @param schedule to check if it should fire
+     * 
+     * @returns true if the given schedule should fire; false otherwise
+     */
+    private fun shouldFire( schedule: EmailSchedule ): Boolean{
+
+        val parser = CronParser( CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX) );
+        // Get date for last execution
+        val now = ZonedDateTime.now();
+        val executionTime = ExecutionTime.forCron(parser.parse(schedule.cronSchedule));
+        val timeFromLastExecution = executionTime.timeFromLastExecution(now);
+
+        return ( timeFromLastExecution.get().toSeconds() <= 4*60 /* 4 minutes */);
```

Comment:
I am not sure I 100% understand (and the cron-utils docs aren't super clear either - I'm looking at https://github.com/jmrozanec/cron-utils/blob/master/src/main/java/com/cronutils/model/time/ExecutionTime.java and https://github.com/jmrozanec/cron-utils/blob/master/src/main/java/com/cronutils/model/time/SingleExecutionTime.java)

Is my understanding correct that...
```kotlin
val timeFromLastExecution = executionTime.timeFromLastExecution(now);
```
...grabs a duration objection that contains how long ago it was since the last time (something  in the past) this schedule would have fired?

If so, then should this `<=` be a `>=` instead?

I guess the second part here is: what is this 4 minute timeout doing here and why is it different from the (default) 5 minute timeout used by the schedule in
```kotlin
/*56:  */ const val SCHEDULE = "*/5 * * * *" // every 5 minutes
/* ... */
/*134: */ fun run(
/*135: */         @TimerTrigger( name = "emailScheduleEngine", schedule = SCHEDULE ) timerInfo : String,
/*136: */         context: ExecutionContext
/*137: */     )
```

I think turning this magic value of 4 minutes into a constant with an explanatory name might clear up some things.

Reply:
So, the timer function doesn't fire on exactly every 5 minutes, so this is the "catch" at assure that anytime within less than 5 minutes of the last timer execution is valid (i.e. timer trigger fires at 11:59 but the cron for the schedule is at noon, without this, they don't "line up" and the schedule is always reported as shouldFire = false)  This accounts for a difference in cron timers.